### DISCUSSION TITLE
Loading third-party dependencies into KC

### DISCUSSION THREAD

**Discussion by mimozell** [keycloak/keycloak#10496] [Archive]
Hi, 

I'm new to Keycloak and we just switched to version 17, but we had the same issue in the previous version. I am working on emitting user change events to AWS Kinesis, but I notice the following errors when running the code:

`java.lang.NoClassDefFoundError: com/amazonaws/auth/AWSCredentialsProvider`. This happens with other classes as well. I am importing the Kinesis dependency using Gradle in the following way:

```
dependencies {
    implementation 'org.keycloak:keycloak-core:17.0.0'
    implementation 'org.keycloak:keycloak-server-spi:17.0.0'
    implementation 'org.keycloak:keycloak-server-spi-private:17.0.0'
    implementation 'org.keycloak:keycloak-services:17.0.0'
    implementation 'org.jetbrains.kotlin:kotlin-stdlib-jdk8'

    implementation 'com.amazonaws:amazon-kinesis-producer:0.14.10'
}
```
`AWSCredentialsProvider` is a dependency under the `implementation 'com.amazonaws:amazon-kinesis-producer:0.14.10'` library  so  I would expect not to have to import anything else in order to support that. I wouldn't have to if this was a Spring Boot application.

In a similar way, I am struggling getting Kotlin support. Ideally I would want it to be covered by `build.gradle` and not to have to set up providers specifically for all the different dependencies I might need or the dependencies those dependencies need. Is there something I am missing during the packaging of the jar that is causing this issue or are we fundamentally expected to provide the jars for each dependency and third-party dependencies manually?

For Wildfly I had to set up a script like this:

```
module add --name=org.jetbrains.kotlin --dependencies=org.jetbrains.annotations,org.jetbrains.kotlin-stdlib-common --resources=kotlin-stdlib.jar
module add --name=org.jetbrains.annotations --resources=kotlin-annotations.jar
module add --name=org.jetbrains.kotlin-stdlib-common --resources=kotlin-stdlib-common.jar
```

and then Kotlin was supported, but not the Kinesis library and its dependencies. Once I added the Kinesis jar, still the dependency that `AWSCredentialsProvider` came from wasn't supported. If we have to add all the jars for all the dependencies separately, that would bring in a lot of work as the third party dependencies can be "endless".

**Comment by mimozell** [comment]
When building a fat jar and adding that as a custom provider, I get the following error:

```
#12 0.689 Updating the configuration and installing your custom providers, if any. Please wait.
#12 7.516 ERROR: Failed to run 'build' command.
#12 7.516 ERROR: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
#12 7.516       [error]: Build step org.keycloak.quarkus.deployment.KeycloakProcessor#configureProviders threw an exception: java.util.ServiceConfigurationError: org.keycloak.provider.Spi: org.keycloak.exportimport.ClientDescriptionConverterSpi not a subtype
#12 7.516       at java.base/java.util.ServiceLoader.fail(ServiceLoader.java:589)
#12 7.516       at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNextService(ServiceLoader.java:1237)
#12 7.516       at java.base/java.util.ServiceLoader$LazyClassPathLookupIterator.hasNext(ServiceLoader.java:1265)
#12 7.516       at java.base/java.util.ServiceLoader$2.hasNext(ServiceLoader.java:1300)
#12 7.516       at java.base/java.util.ServiceLoader$3.hasNext(ServiceLoader.java:1385)
#12 7.516       at org.keycloak.provider.DefaultProviderLoader.loadSpis(DefaultProviderLoader.java:47)
#12 7.516       at org.keycloak.provider.ProviderManager.loadSpis(ProviderManager.java:79)
#12 7.516       at org.keycloak.quarkus.deployment.KeycloakProcessor.loadFactories(KeycloakProcessor.java:454)
#12 7.516       at org.keycloak.quarkus.deployment.KeycloakProcessor.configureProviders(KeycloakProcessor.java:254)
#12 7.516       at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
#12 7.516       at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
#12 7.516       at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
#12 7.516       at java.base/java.lang.reflect.Method.invoke(Method.java:566)
#12 7.516       at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:882)
#12 7.516       at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
#12 7.516       at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
#12 7.516       at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
#12 7.516       at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
#12 7.516       at java.base/java.lang.Thread.run(Thread.java:829)
#12 7.516       at org.jboss.threads.JBossThread.run(JBossThread.java:501)
#12 7.516 
```

**Comment by stianst** [comment]
Please check the server developer guide. In essence just copy additional dependencies to the providers directory

**Comment by mimozell** [comment]
This is what the dependency tree looks like for the Kinesis library:
```
\--- com.amazonaws:amazon-kinesis-producer:0.14.10
     +--- com.google.guava:guava:29.0-jre -> 30.1-jre (*)
     +--- org.slf4j:slf4j-api:1.7.25 -> 1.7.30
     +--- com.google.protobuf:protobuf-java:3.11.4
     +--- commons-io:commons-io:2.7
     +--- commons-lang:commons-lang:2.6
     +--- com.amazonaws:aws-java-sdk-core:1.11.960
     |    +--- commons-logging:commons-logging:1.1.3 -> 1.2
     |    +--- org.apache.httpcomponents:httpclient:4.5.13 (*)
     |    +--- software.amazon.ion:ion-java:1.0.2
     |    +--- com.fasterxml.jackson.core:jackson-databind:2.6.7.4 -> 2.12.2 (*)
     |    +--- com.fasterxml.jackson.dataformat:jackson-dataformat-cbor:2.6.7 -> 2.12.2
     |    |    +--- com.fasterxml.jackson.core:jackson-databind:2.12.2 (*)
     |    |    +--- com.fasterxml.jackson.core:jackson-core:2.12.2
     |    |    \--- com.fasterxml.jackson:jackson-bom:2.12.2 (*)
     |    \--- joda-time:joda-time:2.8.1 -> 2.10.2
     +--- software.amazon.glue:schema-registry-serde:1.1.5
     |    +--- software.amazon.glue:schema-registry-common:1.1.5
     |    |    +--- software.amazon.awssdk:glue:2.17.52
     |    |    |    +--- software.amazon.awssdk:aws-json-protocol:2.17.52
     |    |    |    |    +--- software.amazon.awssdk:aws-core:2.17.52
     |    |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    +--- software.amazon.awssdk:regions:2.17.52
     |    |    |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    |    +--- software.amazon.awssdk:utils:2.17.52
     |    |    |    |    |    |    |    +--- org.reactivestreams:reactive-streams:1.0.3
     |    |    |    |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    |    |    \--- org.slf4j:slf4j-api:1.7.30
     |    |    |    |    |    |    +--- software.amazon.awssdk:sdk-core:2.17.52
     |    |    |    |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    |    |    +--- software.amazon.awssdk:http-client-spi:2.17.52
     |    |    |    |    |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    |    |    |    |    |    +--- software.amazon.awssdk:metrics-spi:2.17.52
     |    |    |    |    |    |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    |    |    |    |    \--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    |    |    |    |    |    \--- org.reactivestreams:reactive-streams:1.0.3
     |    |    |    |    |    |    |    +--- software.amazon.awssdk:metrics-spi:2.17.52 (*)
     |    |    |    |    |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    |    |    |    |    +--- software.amazon.awssdk:profiles:2.17.52
     |    |    |    |    |    |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    |    |    |    |    |    \--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    |    |    +--- org.slf4j:slf4j-api:1.7.30
     |    |    |    |    |    |    |    \--- org.reactivestreams:reactive-streams:1.0.3
     |    |    |    |    |    |    +--- software.amazon.awssdk:profiles:2.17.52 (*)
     |    |    |    |    |    |    +--- software.amazon.awssdk:json-utils:2.17.52
     |    |    |    |    |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    |    |    \--- software.amazon.awssdk:third-party-jackson-core:2.17.52
     |    |    |    |    |    |    \--- org.slf4j:slf4j-api:1.7.30
     |    |    |    |    |    +--- software.amazon.awssdk:auth:2.17.52
     |    |    |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    |    |    |    +--- software.amazon.awssdk:sdk-core:2.17.52 (*)
     |    |    |    |    |    |    +--- software.amazon.awssdk:regions:2.17.52 (*)
     |    |    |    |    |    |    +--- software.amazon.awssdk:profiles:2.17.52 (*)
     |    |    |    |    |    |    +--- software.amazon.awssdk:http-client-spi:2.17.52 (*)
     |    |    |    |    |    |    +--- software.amazon.awssdk:json-utils:2.17.52 (*)
     |    |    |    |    |    |    \--- software.amazon.eventstream:eventstream:1.0.1
     |    |    |    |    |    +--- software.amazon.awssdk:profiles:2.17.52 (*)
     |    |    |    |    |    +--- software.amazon.awssdk:sdk-core:2.17.52 (*)
     |    |    |    |    |    +--- software.amazon.awssdk:http-client-spi:2.17.52 (*)
     |    |    |    |    |    +--- software.amazon.awssdk:metrics-spi:2.17.52 (*)
     |    |    |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    |    |    \--- software.amazon.eventstream:eventstream:1.0.1
     |    |    |    |    +--- software.amazon.awssdk:protocol-core:2.17.52
     |    |    |    |    |    +--- software.amazon.awssdk:sdk-core:2.17.52 (*)
     |    |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    |    |    \--- software.amazon.awssdk:http-client-spi:2.17.52 (*)
     |    |    |    |    +--- software.amazon.awssdk:sdk-core:2.17.52 (*)
     |    |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    |    +--- software.amazon.awssdk:http-client-spi:2.17.52 (*)
     |    |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    |    +--- software.amazon.awssdk:third-party-jackson-core:2.17.52
     |    |    |    |    \--- software.amazon.awssdk:json-utils:2.17.52 (*)
     |    |    |    +--- software.amazon.awssdk:protocol-core:2.17.52 (*)
     |    |    |    +--- software.amazon.awssdk:sdk-core:2.17.52 (*)
     |    |    |    +--- software.amazon.awssdk:auth:2.17.52 (*)
     |    |    |    +--- software.amazon.awssdk:http-client-spi:2.17.52 (*)
     |    |    |    +--- software.amazon.awssdk:regions:2.17.52 (*)
     |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    +--- software.amazon.awssdk:aws-core:2.17.52 (*)
     |    |    |    \--- software.amazon.awssdk:metrics-spi:2.17.52 (*)
     |    |    +--- software.amazon.glue:schema-registry-build-tools:1.1.5
     |    |    +--- software.amazon.awssdk:aws-json-protocol:2.17.52 (*)
     |    |    +--- software.amazon.awssdk:url-connection-client:2.17.52
     |    |    |    +--- software.amazon.awssdk:utils:2.17.52 (*)
     |    |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    |    \--- software.amazon.awssdk:http-client-spi:2.17.52 (*)
     |    |    +--- org.apache.avro:avro:1.10.2
     |    |    |    +--- com.fasterxml.jackson.core:jackson-core:2.12.2
     |    |    |    +--- com.fasterxml.jackson.core:jackson-databind:2.12.2 (*)
     |    |    |    \--- org.slf4j:slf4j-api:1.7.30
     |    |    +--- org.apache.commons:commons-compress:1.21
     |    |    +--- org.slf4j:slf4j-api:1.7.30
     |    |    +--- com.google.guava:guava:30.0-jre -> 30.1-jre (*)
     |    |    +--- org.apache.commons:commons-lang3:3.8.1
     |    |    +--- org.projectlombok:lombok:1.18.20
     |    |    \--- org.projectlombok:lombok-utils:1.18.12
     |    +--- software.amazon.awssdk:arns:2.17.52
     |    |    +--- software.amazon.awssdk:annotations:2.17.52
     |    |    \--- software.amazon.awssdk:utils:2.17.52 (*)
     |    +--- org.apache.kafka:kafka-clients:2.8.1
     |    |    +--- com.github.luben:zstd-jni:1.4.9-1
     |    |    +--- org.lz4:lz4-java:1.7.1
     |    |    +--- org.xerial.snappy:snappy-java:1.1.8.1
     |    |    \--- org.slf4j:slf4j-api:1.7.30
     |    +--- com.kjetland:mbknor-jackson-jsonschema_2.12:1.0.39
     |    |    +--- org.scala-lang:scala-library:2.12.10
     |    |    +--- org.jetbrains.kotlin:kotlin-scripting-compiler-embeddable:1.3.50
     |    |    +--- com.fasterxml.jackson.core:jackson-databind:2.10.1 -> 2.12.2 (*)
     |    |    +--- javax.validation:validation-api:2.0.1.Final
     |    |    \--- org.slf4j:slf4j-api:1.7.26 -> 1.7.30
     |    +--- io.github.classgraph:classgraph:4.8.120
     |    \--- com.github.erosb:everit-json-schema:1.12.2
     |         +--- org.json:json:20201115
     |         +--- commons-validator:commons-validator:1.6
     |         |    +--- commons-digester:commons-digester:1.8.1
     |         |    +--- commons-logging:commons-logging:1.2
     |         |    \--- commons-collections:commons-collections:3.2.2
     |         +--- com.damnhandy:handy-uri-templates:2.1.8
     |         |    \--- joda-time:joda-time:2.10.2
     |         \--- com.google.re2j:re2j:1.3
     \--- jakarta.xml.bind:jakarta.xml.bind-api:2.3.2
          \--- jakarta.activation:jakarta.activation-api:1.2.1
```
are people really expected to find the relevant dependencies for their usage of the library and then upload the jars for each of those manually? Doesn't seem like a practical solution at all.

**Reply by yiumen** [reply]
If using maven, use the maven assembly plugin to build a zip / tar archive that includes the built jar and all its runtime dependencies.  Use dependencySets in assembly descriptor. see https://maven.apache.org/plugins/maven-assembly-plugin/assembly.html.

If using gradle, use gradle Java Library Distribution Plugin. see https://docs.gradle.org/current/userguide/java_library_distribution_plugin.html

**Comment by mimozell** [comment]
@stianst I found this similar issue to mine with the fat jar from older versions: https://issues.redhat.com/browse/KEYCLOAK-4428?workflowName=GIT+Pull+Request+workflow+&stepId=6. Here the solution was to include keycloak-services in the `jboss-deployment-structure.xml`. What is the equivalent now that that has been removed?

**Reply by stianst** [reply]
You don't need to include keycloak-services in your fatjar/jar/etc on Quarkus dist as there is a single classpath, so your provider will have access to everything already.

**Reply by mimozell** [reply]
Okay, I excluded the keycloak-specific dependencies and added all the other jars through gradle. But I get the following error:
`javax.validation.NoProviderFoundException: Unable to create a Configuration, because no Jakarta Bean Validation provider could be found. Add a provider like Hibernate Validator (RI) to your classpath.` I would expect it to be in the classpath since it is a KC dependency. 
I tried adding the following:
```
    implementation 'jakarta.validation:jakarta.validation-api:2.0.2'
    implementation 'org.glassfish:jakarta.el:4.0.2'
    implementation 'org.hibernate:hibernate-validator:7.0.3.Final'
```
and their jars to the providers, but it changed nothing. Which validator is expected and should I be including it or should it be there?

**Reply by pedroigor** [reply]
In order to enable bean validation, we should be introducing the `quarkus-hibernate-validator` extension. The dependencies we have in the classpath are only the API/SPI and the extension is needed to actually enable the capability.

Are you able to share your provide code or provide a simple example that I can work with in order to evaluate if introducing the extension is enough?

**Reply by mimozell** [reply]
Yes, I should be able to provide a simple example within the week. Thanks!

**Reply by mimozell** [reply]
@pedroigor I created the simplest setup I could here: https://github.com/mimozell/keycloak-extensions. To reproduce the issue, you just log in (admin:admin), set the custom event listener and start clicking around to get a stacktrace. I realize that I actually have the validation jar that is needed as part of the custom providers, but somehow it is not recognized by the Kinesis library as being in the classpath. I'm not sure if this is a KC issue or something else.

**Reply by pedroigor** [reply]
@mimozell Thanks.

After adding the hibernate validator extension to the distribution the errors are gone and it seems to work fine (even though not yet integrating with Kinesis).

Of course, I've also added all jars from `build/libs/jars` to the `providers` directory in addition to the provider itself.

Would you be able to build the distribution from this branch [1] and then check how it works for you?

This should not be the final solution though and the proper solution involves solving a more general issue when deploying providers with 3rd party dependencies. But I would like to see at least how it works for you to make sure we have something to think/build on top.

[1] https://github.com/pedroigor/keycloak/tree/tmp-add-hhh-validator

**Reply by pedroigor** [reply]
I'm going to try to slim down too the dependencies that you actually need to include into providers directory so that we can also think how to make easier the step of figuring out which dependencies you need to include.

**Reply by mimozell** [reply]
How are you building the docker image out of the source? I didn't find any documentation around it or a relevant Dockerfile. I tried:
`docker build --build-arg GIT_REPO=pedroigor/keycloak/ --build-arg GIT_BRANCH=tmp-add-hhh-validator .` as reported here: https://hub.docker.com/r/jboss/keycloak/, but it didn't work.

**Reply by pedroigor** [reply]
I'm not building an image but running locally.

Here is the branch https://github.com/pedroigor/keycloak/tree/tmp-add-hhh-validator.

**Reply by mimozell** [reply]
Since I am deploying a container it would be great if there could be a guide on how to build the Docker image out of your branch and then I just keep things exactly as they are on my side. If not, then pointing to some guide on how to start the service locally and how to easily extend it with my jar would be great.

**Comment by pedroigor** [comment]
@DGuhr @stianst I think we have a pretty good use case to start thinking about how to deal with providers using 3rd party dependencies. IMO, if we can think/build something on top of it we should be able to solve most of the issues we are facing now when migrating providers from Wildfly:

* How developers know which dependencies we provide OOTB
* How to package/deploy providers and their 3rd party dependencies (copy both providers and all 3rd party is not the best UX)
* How to deal with providers using different library versions

**Reply by dmitrygusev** [reply]
hi @pedroigor, do you know if any work was done on the points from your list above?

I have a related question: is it possible to package providers as JPMS/Jigsaw modules?

I assume not at the moment because of the [single classpath](https://github.com/keycloak/keycloak/discussions/10496#discussioncomment-2278515), but did you consider this/is it something that may appear in the nearest future?

Thanks

**Reply by dmitrygusev** [reply]
It looks like in theory, it is possible to use both module path and classpath in the same app [1].

So, again, in theory, it should be possible to create a provider JAR as a JPMS module and add it explicitly to the Keycloak classpath using the `--add-modules` option, unless I miss something.

Related article on java service loaders with JPMS:
https://www.infoq.com/articles/java11-aware-service-module/

[1] https://stackoverflow.com/a/46289257/2414933

**Reply by pedroigor** [reply]
TBH, I did not try it yet but as you said, it should be possible in theory.

However, whatever JAR you put into `providers` directory it will be loaded by the application class loader. I'm wondering then if using JPMS makes sense.

**Comment by mimozell** [comment]
In `build.gradle` I ended up doing:

```
task copyToLib(type: Copy) {
    from configurations.runtimeClasspath.exclude(group: 'org.keycloak')
    into "$buildDir/libs/jars"
    duplicatesStrategy "exclude"
}
jar { dependsOn copyToLib }
```
which copies the jars of the libraries that I am referencing and their dependencies into the jars directory excluding everything by keycloak which should already be present (`group: 'org.keycloak'`), and then I copy those through the Dockerfile into KC like this:

```
COPY build/libs/jars /opt/keycloak/providers/
```

**Comment by jesevis** [comment]
If there's still someone facing the same Keycloak dependency issue and doesn't know how to go forward, I’ve put together a guide that's related to the helpful advice discussed here. Mainly because I didn't find such a guide myself. It covers the steps required to make 3rd party dependencies available inside Keycloak from scratch. At least it can be a starting point to build an even better solution on top.

https://medium.com/@jesevis/using-3rd-party-dependencies-in-keycloak-a63208d43cba

**Comment by mabartos** [comment]
Right now, it is quite hard to provide an external dependency within your extension without the shaded JARs.

In the meantime, there is this [Keycloak Quarkus extension](https://github.com/mabartos/keycloak-quarkus-extensions) community project, which will build your own Keycloak distribution based on the upstream build mechanism with the required dependencies or Quarkus extensions. You can check the [examples](https://github.com/mabartos/keycloak-quarkus-extensions/tree/main/examples#readme) on how to achieve this.

For instance, you should be able to add the dependency via the tool as follows:
```bash
# add the required dependency
./kc-extension.sh add com.amazonaws:amazon-kinesis-producer:0.14.10

# build the custom distribution
./kc-extension.sh build

# start the custom Keycloak instance and try it out
./kc-extension.sh start-dev 

# (optional) create a builder container image
./kc-extension.sh image
```
